docs/SubmittingChanges.md

```diff
@@ -0,0 +1,159 @@
+
+* `git gui` to interactively commit only a subset of local changes, and to
+  amend the most recent commit.
+* `git add -p` to achieve the same via the command line.
+* `git commit --amend` to amend the most recent commit from the command line.
+* `git stash` to temporarily undo a change and re-apply it later.
+* `git commit --squash=<commit>` and `git commit --fixup=<commit>` to mark
+  changes to be added to an earlier commit.
+* `git rebase -i <commit>` and the `rebase.autoSquash` option to rework a
+  series of commits.
+* `git rebase -i <commit> -x "..."` to execute a command after each commit;
+  you can use this to verify that LLPC compiles and passes lit tests after
+  each commit.
+* `git cherry-pick` for moving a commit from one branch to another.
+* `gitk` and `tig` to browse the repository's history.
+* `git reflog` to find old states of the repository that may seem lost; once
+  you commit something, Git will **not** lose it -- at least not for the next
+  30 days.
+
+Similar functionality is available in some GUI tools. Learning the tools of the
+trade makes you a better developer.
+
+Having multiple incomplete commits locally and rewriting them using
+`git rebase` is normal and encouraged. The guidelines above only apply for
+sending out your changes via a pull request.
+
+
+## Write useful commit messages
+
+[Parts of this section are also adapted from
+[the Linux kernel guide on submitting patches](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/Documentation/process/submitting-patches.rst).]
+
+Commit messages should describe the "why" of a change: What does the change
+solve? Where does it lead? Convince the reviewer that there is a problem worth
+fixing.
+
+Follow best practices for commit messages:
+
+* The first line must be a short summary of the commit. The second line must be
+  blank.
+* When the commit is contained entirely within a logical module, prefix
+  the summary with the module name, e.g. "builder: When the summary
+* Generally keep lines shorter than 75 characters. Exceeding this limit
+  slightly for the summary is okay, and don't break lines where it would
+  hurt searchability, e.g. don't break URLs and CTS test names. Don't break
+  footer lines, as tools should be able to parse them mechanically.
+* Use common practice footer lines to provide more information where
+  applicable.
+* The most important footer line is a "Fixes:" line. They allow a quick
+  identification of relevant bug fix commits, for example for cherry-picking
+  onto release branches. Common forms are:
+  * Indicate the fix of a regression that was introduced by an earlier commit,
+    using both the hash and the summary title of the commit in the following
+    format:
+
+        Fixes: 123456789a ("patch: Change the default order of userdata SGPRs")
+
+  * Refer to an issue that was fixed:
+  
+        Fixes: https://github.com/GPUOpen-Drivers/llpc/issues/1234
```

Comment:
The format GitHub uses would be `Fixes: #1234`, which is linked to the GitHub issues and automatically closes the issue, once a commit with this message is merged.

The downside of these (GitHub-)references in commit messages is, if the commit is rebased often, the issue on GitHub gets spammed with references to all these commits.

Reply:
I've just tested this -- GitHub also understands the URL form -- see here: https://github.com/nhaehnle/llpc/issues/1 -- and I find the URL preferable because it's more explicit and allows simply copy&paste or even direct clicks on the link from most tools (e.g. even from the terminal window)